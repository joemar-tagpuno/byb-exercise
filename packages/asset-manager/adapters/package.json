{
  "private": false,
  "name": "@byb/asset-manager-adapters",
  "version": "0.1.0",
  "files": ["build"],
  "main": "./build/index.js",
  "module": "./build/index.mjs",
  "types": "./build/index.d.ts",
  "__exports-comment": "// Define output files for Node.js export maps (https://nodejs.org/api/packages.html#exports)",
  "exports": {
    ".": {
      "import": {
        "@byb/source": "./src/index.ts",
        "types": "./build/index.d.mts",
        "import": "./build/index.mjs",
        "default": "./build/index.mjs"
      },
      "require": {
        "@byb/source": "./src/index.ts",
        "types": "./build/index.d.ts",
        "import": "./build/index.js",
        "default": "./build/index.js"
      }
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "clean": "rimraf build .turbo",
    "build": "tsup ./src/index.ts --tsconfig ./tsconfig.build.json --config ../../../tsup.config.ts",
    "build:watch": "pnpm build --watch",
    "check:exports": "attw --pack .",
    "test": "vitest --workspace ../../../vitest.workspace.ts run",
    "test:unit": "vitest --workspace ../../../vitest.workspace.ts --project unit run",
    "test:integration": "vitest --workspace ../../../vitest.workspace.ts --project integration run"
  },
  "dependencies": {
    "@byb/asset-manager-core": "workspace:*",
    "inversify": "^6.2.1",
    "reflect-metadata": "^0.2.2",
    "sequelize": "^6.37.5"
  },
  "devDependencies": {
    "nock": "^13.5.6",
    "sqlite3": "^5.1.7"
  }
}
