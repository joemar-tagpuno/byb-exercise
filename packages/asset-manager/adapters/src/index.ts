export { AssetManagerModule } from './AssetManagerModule'
export { DownloadableFileLoaderGateway } from './gateways/file-download/DownloadableFileLoaderGateway'
export { InMemoryDownloadableFileRepository } from './repositories/file-download/InMemoryDownloadableFileRepository'
