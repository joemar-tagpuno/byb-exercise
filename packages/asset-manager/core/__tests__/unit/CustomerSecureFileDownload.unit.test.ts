import type { StubbedType } from '@byb/testing'
import { beforeEach, describe, expect, it } from 'vitest'
import {
  AssetManagerErrors,
  CustomerSecureFileDownload,
  type DownloadableFileLoader,
  type DownloadableFileRepository,
} from '../../src'
import { fakeDownloadableFileLoader, fakeDownloadableFileRepository } from '../adapters/Mocks'

describe('CustomerSecureFileDownload', () => {
  let downloadableFileRepository: StubbedType<DownloadableFileRepository>
  let downloadableFileLoader: StubbedType<DownloadableFileLoader>
  let useCase: CustomerSecureFileDownload

  beforeEach(() => {
    downloadableFileRepository = fakeDownloadableFileRepository
    downloadableFileLoader = fakeDownloadableFileLoader
    useCase = new CustomerSecureFileDownload(downloadableFileRepository, downloadableFileLoader)
  })

  describe('Everything is ok', () => {
    it('can guard the secure download use-case', async () => {
      // GIVEN
      const identity = {
        id: 'unknown-id',
        userType: 'unknown',
        userRole: 'unknown',
        email: 'email@fake.com',
        phone: '0912345678',
      }
      const request = {
        userId: 'user-id',
        fileId: 'file-id',
      }

      // WHEN
      const result = await useCase.canExecute(identity, request)

      // THEN
      expect(result).toBeFalsy()
      expect(result).toMatchSnapshot()
    })

    it('can execute the secure download use-case', async () => {
      // GIVEN
      downloadableFileRepository.findById.mockResolvedValue({
        name: 'file-name.pdf',
        url: 'http://localhost:123456/file-url',
        size: 124,
        type: 'pdf',
        provider: 'http',
      })

      downloadableFileLoader.load.mockResolvedValue(Buffer.from('file-content-base64'))

      const input = {
        userId: 'user-id',
        fileId: 'file-id',
      }

      // WHEN
      const result = await useCase.execute(input)

      // THEN
      expect(result).toEqual({
        file: Buffer.from('file-content-base64').toString('base64'),
        filename: 'file-name.pdf',
        filesize: 124,
        filetype: 'pdf',
      })
      expect(result).toMatchSnapshot()
    })
  })

  // Can fail when
  it('can throw a domain error if the file cannot be found', async () => {
    // GIVEN
    downloadableFileRepository.findById.mockRejectedValueOnce(
      new AssetManagerErrors.FileDownload.DownloadableFileModelNotFound(),
    )
    const input = {
      userId: 'user-id',
      fileId: 'file-id',
    }

    // WHEN
    const result = useCase.execute(input)

    // THEN
    await expect(result).rejects.toThrow(AssetManagerErrors.FileDownload.DownloadableFileModelNotFound)
  })
})
