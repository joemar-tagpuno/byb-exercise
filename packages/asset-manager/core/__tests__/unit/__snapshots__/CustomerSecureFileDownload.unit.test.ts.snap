// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomerSecureFileDownload > Everything is ok > can execute the secure download use-case 1`] = `
{
  "file": "ZmlsZS1jb250ZW50LWJhc2U2NA==",
  "filename": "file-name.pdf",
  "filesize": 124,
  "filetype": "pdf",
}
`;

exports[`CustomerSecureFileDownload > Everything is ok > can guard the secure download use-case 1`] = `false`;
